---
title: "R code for Figure 4 - figure supplement 1 - Analysis of ATAC-seq DNA accessibility in cerebellar granule and basket cells."
author: "Xiao Xu (xxu@rockefeller.edu)"
date: "August 2018"
output: 
  html_document:
    toc: yes
  html_notebook:
    toc: yes
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. This notebook contains R code for figures and analyses related to Figure 4 - figure supplement 1 in the manuscript:

Xu X, Stoyanova EI, Lemiesz A, Xing J, Mash DC, Heintz N (2018). Species and Cell-Type Properties of Classically Defined Human and Rodent Neurons and Glia.

Specifically, this workflow will generate the figures related to analysis of cell-type specific ATAC peaks in Figure 4 - figure supplement 1.

Load libraries and functions:

```{r warning = FALSE, message = FALSE}
# source custom functions
source("functions.R")

# libraries for gene expression analysis
library(DESeq2)

# libraries for colors
library(RColorBrewer)

# libraries for data manipulation
library(reshape2)

# work in the tidyverse
library(tidyverse)
```


## Peak mapping

Produce a stacked barchart showing for different samples where peaks map in the genome.

Load data on peak mapping.

```{r message = FALSE}
# import table with peak summary
peaks_sum <- read_tsv("../data/atac_peak_summary.txt")

# lookup table for condensing annotation
regions <- c("Promoter", "GB", "Intergenic", "Other")
lookup <- rep(regions, c(1, 5, 1, 5))
names(lookup) <- c("Promoter", "5UTR", "Exon", "Intron", "3UTR", "TTS",
                   "Intergenic", "pseudo", "miRNA", "ncRNA", "snoRNA", "rRNA")
# specify sample level for plotting
sample_levels <- c(
  "human_granule_all", "human_granule_da", "human_basket_all", "human_basket_da",
  "mouse_granule_all", "mouse_granule_da", "mouse_basket_all", "mouse_basket_da"
  )

# 
ggdata <- peaks_sum %>%
  mutate(Sample = paste(Species, Cell_type, Type, sep = "_")) %>%
  rename(Region_detail = Region) %>%
  mutate(Region = lookup[Region_detail]) %>%
  mutate(Region = factor(Region, levels = regions)) %>%
  mutate(Sample = factor(Sample, levels = sample_levels))
```


Summarize peaks by genomic region

```{r}
ggdata %>%
  group_by(Sample, Region) %>%
  summarise(Peaks = sum(Peaks)) %>%
  spread(key = Region, value = Peaks) %>%
  mutate(Total = sum(Other, Intergenic, GB, Promoter))
```


Stacked barplot

```{r}
ggplot(ggdata, aes(x = Sample, y = Peaks, fill = Region, label = Peaks)) +
  geom_bar(position = "fill", stat = "identity") +
  scale_fill_brewer(type = "qual", palette = "Set2") +
  scale_x_discrete(limits = rev(levels(ggdata$Sample))) +
  coord_flip() +
  theme_bw()
ggsaveToPdf("../figures/figure_4_s1_atac_peaks_by_region.pdf")
```


## Granule vs Basket cell DE genes

Find differentially expressed genes between granule cells and basket cells and output for ATAC promoter coverage analysis.

Set cutoffs and define helper function for summarizing and saving de genes to disk

```{r}
p_cutoff <- 10e-5
bm_cutoff <- 50

getSumTableAndOutput <- function(de_table, p, bm, filename) {
  write_tsv(
    de_table %>% 
      filter(padj < p, baseMean > bm) %>% 
      arrange(log2FoldChange),
    paste("../output/", filename, sep = "")
  )
  
  de_table %>%
    filter(baseMean > bm, padj < p) %>%
    group_by(log2FoldChange > 0) %>%
    count()
}
```


Mouse

```{r message = FALSE, warning = FALSE}
# import counts and define metadata
counts <- read_tsv("../data/mouse_gfp_antibody_wg_counts.txt") %>%
  select(Gene, contains("neun_pos"), contains("sorcs3"))
meta <- tibble(
  Sample = colnames(counts)[-1],
  Cell_type = c( rep("Granule", 3), rep("Basket", 3) )
  )

# perform de, get summary table, output to file
de_table <- getDECustom(counts, meta, "Granule", "Basket")
getSumTableAndOutput(de_table, p_cutoff, bm_cutoff, 
                     "figure_4_s1_de_mouse_granule_vs_basket.txt")
```

Human

```{r message = FALSE, warning = FALSE}
# import counts and define metadata
counts <- read_tsv("../data/human_xk_pk_wg_counts.txt") %>%
  select(Gene, contains("granule"), contains("basket"))
meta <- tibble(
  Sample = colnames(counts)[-1],
  Cell_type = c( rep("Granule", 2), rep("Basket", 2) )
  )

# perform de, get summary table, output to file
de_table <- getDECustom(counts, meta, "Granule", "Basket")
getSumTableAndOutput(de_table, p_cutoff, bm_cutoff, 
                     "figure_4_s1_de_human_granule_vs_basket.txt")
```




## Plot expression of genes associated with ATAC DA peaks

Import peaks. Keep peaks that map to promoters or gene bodies of protein coding genes only. Condense annotation.

```{r message = FALSE, warning = FALSE}
# lookup table for condensing annotation
regions <- c("Promoter", "5UTR", "Exon_intron", "3UTR_TTS")
lookup <- rep(regions, c(1, 1, 3, 2))
names(lookup) <- c("promoter-TSS", "5'", "exon", "intron", "non-coding", "3'", "TTS")

# import peaks and condense annotation
peaks <- read_tsv("../data/atac_da_peaks.txt") %>%
  separate(Sample, into = c("Species", "Cell_type"), sep = "_", remove = FALSE) %>%
  filter(Gene.Type == "protein-coding", Annotation != "Intergenic") %>% 
  separate(Annotation, into = c("Region_detail", "Notes"), sep = " ", remove = FALSE) %>%
  mutate(Region = lookup[Region_detail])
```


Load granule and basket normalized counts from mouse and human
```{r message = FALSE}
norm_exp <- list()
mouse_file <- "../data_clean/figure_1_main_mouse_gfp_antibody_norm_rld.RDS"
human_file <- "../data_clean/figure_2_main_human_norm_rld.RDS"

# mouse
norm_exp[["mouse"]] <- read_rds(mouse_file) %>%
  assay() %>%
  as.data.frame() %>%
  rownames_to_column("Gene") %>%
  select(Gene, contains("neun_pos"), contains("sorcs3"))
colnames( norm_exp[["mouse"]] ) <- c("Gene", "Granule.1", "Granule.2", "Granule.3",
                                     "Basket.1", "Basket.2", "Basket.3")

# human
norm_exp[["human"]] <- read_rds(human_file) %>%
  assay() %>%
  as.data.frame() %>%
  rownames_to_column("Gene") %>%
  select(Gene, contains("xk"), contains("pk")) %>%
  select(Gene, contains("granule"), contains("basket"))
colnames( norm_exp[["human"]] ) <- c("Gene", "Granule.1", "Granule.2", "Basket.1", "Basket.2")
```


Plot distribution of DA genes, save plot to disk, get statistics for difference between cell types

```{r message = FALSE, warning = FALSE}
species <- c("human", "mouse")
cell_types <- c("granule", "basket")
lapply(
  species,
  function(s) lapply(
    cell_types,
    function(ct) {
      # subset peaks
      peaks_sub <- peaks %>%
        filter(Species == s) %>%
        filter(Cell_type == ct) %>%
        select(Gene =  Gene.Name, Region) %>%
        distinct(Gene, Region, .keep_all = TRUE) %>%
        inner_join(norm_exp[[s]], by = "Gene")
      
      # melt for plotting
      exp_melt <- peaks_sub %>%
        melt() %>%
        mutate(Region = factor(Region, levels = regions)) %>%
        separate(variable, into = c("Cell_type", "Rep"), sep = "[.]", remove = FALSE)
      
      # boxplot
      condition <- paste(s, ct, sep = "_")
      p <- ggplot(exp_melt, aes(x = variable, y = value)) +
        geom_boxplot(aes(fill = Cell_type)) +
        facet_grid(~ Region) +
        scale_fill_manual(values = c("#FF6600", "#33A02C")) +
        scale_y_continuous(limits = c(0, 17)) +
        ggtitle(condition) +
        theme_bw() + 
        theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1))
      plot(p)
      ggsaveToPdf(paste("../figures/figure_4_s1_boxplot_da_peaks_",
                        condition, ".pdf", sep = ""))
      
      # t-test to determine whether difference in expression is significantly different
      exp_paired <- exp_melt %>%
        separate(variable, into = c("Cell_type", "Rep"), sep = "[.]", remove = FALSE) %>%
        mutate(Gene_region = paste(Gene, Region, Rep, sep = ".")) %>%
        select(Gene_region, Cell_type, value) %>%
        spread(Cell_type, value) %>%
        separate(Gene_region, into = c("Gene", "Region", "Rep"), sep = "[.]")
      
      print(condition)
      sum_stats <- sapply(
        regions,
        function(r) {
          exp_sub <- exp_paired %>% filter(Region == r)
          t <- t.test(exp_sub$Granule, exp_sub$Basket)$p.value
          diff_med <- median(exp_sub$Granule) - median(exp_sub$Basket)
          diff_mean <- mean(exp_sub$Granule) - mean(exp_sub$Basket)
          c(t, diff_med, diff_mean)
        }, USE.NAMES = TRUE)
      rownames(sum_stats) <- c("p_value", "median_diff", "mean_diff")
      
      return(sum_stats)
    }
  )
)
```





