---
title: "R code for Figure 3 - figure supplement 3 - expression of species-enriched genes in published single nuclei data"
author: "Xiao Xu (xxu@rockefeller.edu)"
date: "June 2018"
output: 
  html_document:
    toc: yes
  html_notebook:
    toc: yes
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. This notebook contains R code for figures and analyses related to Figure 3 - figure supplement 3 in the manuscript:

Xu X, Stoyanova EI, Lemiesz A, Xing J, Mash DC, Heintz N (2018). Species and Cell-Type Properties of Classically Defined Human and Rodent Neurons and Glia.

Specifically, this workflow will generate figures relating to the expression of housekeeping or species-enriched genes in human cerebellar single nuclei RNA-seq data from the publication: 

Lake BB, Chen S, Sos BC, Fan J et al. Integrative single-cell analysis of transcriptional and epigenetic states in the human adult brain. Nat Biotechnol 2018 Jan;36(1):70-80. PMID: 29227469

and mouse cerebellar single cell RNA-seq data from the publication:

Saunders A, Macosko EZ, Wysoker A, Goldman M et al. Molecular Diversity and Specializations among the Cells of the Adult Mouse Brain. Cell. 2018 Aug 9;174(4):1015-1030.e16. PMID: 30096299


Load libraries and functions:

```{r warning = FALSE, message = FALSE}
# source custom functions
source("functions.R")

# libraries for gene expression analysis
library(DESeq2)
library(Seurat)
library(DropSeq.util)

# libraries for data import and manipulation
library(data.table)
library(plyr)
library(reshape2)

# work in the tidyverse
library(tidyverse)
```




## Preprare gene lists

Create a dataframe that contains species-enriched and housekeeping genes.

### Import species-enriched genes

```{r message = FALSE}
cell_types <- c("Granule", "Basket", "Astrocyte", "Oligo", "OPC")
se_genes <- lapply(cell_types,
                   function(x) 
                     read_tsv(
                       paste("../output/figure_3_main_species_de_mouse_human_", 
                             tolower(x), ".txt", sep = "")
                       ) %>%
                     mutate(Gene_human = convertIDs(Gene)) %>%
                     mutate(Type = ifelse(log2FoldChange > 0, "Mouse", "Human")) %>%
                     mutate(Cell_type = x) %>%
                     select(Gene = Gene_human, Type, Cell_type)
                   ) %>%
  bind_rows()
```


### Identify housekeeping genes


Read in counts and define metadata

```{r message = FALSE}
counts <- read_tsv("../data/merge_mm_hg_ortho_mouse_human_wg_counts.txt")

meta <- tibble(
  Sample = colnames(counts)[-1],
  Species = c(rep("mouse", 12), rep("human", 10)),
  Cell_type = gsub("^.*_", "", colnames(counts)[-1])
)
```


Define "housekeeping" genes - genes that are unchanged between mouse and human across all cell types.

Run DESeq to find the genes that are not significantly differentially expressed between mouse and human.

```{r message = FALSE, warning = FALSE}
dds <- DESeqDataSetFromMatrix(counts %>% column_to_rownames("Gene"), 
                              meta %>% column_to_rownames("Sample"),
                              design = ~ Cell_type + Species)
dds <- DESeq(dds)

# species comparison
de <- results(dds, contrast = c("Species", "mouse", "human")) %>%
  as.data.frame() %>%
  rownames_to_column("Gene")
```


Calculate fpkms (rpkms) for filtering.

```{r message = FALSE, warning = FALSE}
# read in gene lengths for mouse and human
lengths_mouse <- read_tsv("../data/ref/ensembl_ortho_mh_mouse_gene_longest.txt", 
                          col_names =  c("Gene", "Length") )
# human ids need to be converted to mouse equivalent
lengths_human <- read_tsv("../data/ref/ensembl_ortho_mh_human_gene_longest.txt",
                          col_names =  c("Gene", "Length") ) %>%
  mutate(Gene = convertIDs(Gene, from = "human", to = "mouse"))

# calculate rpkms for mouse and human samples separately
rpkms_mouse <- getNormCounts(counts %>% 
                               select(Gene, 
                                      meta %>% 
                                        filter(Species == "mouse") %>% 
                                        pull(Sample)),
                             lengths_mouse)
rpkms_human <- getNormCounts(counts %>% 
                               select(Gene, 
                                      meta %>% 
                                        filter(Species == "human") %>% 
                                        pull(Sample)),
                             lengths_human)

# merge two tables and get minimum sample rpkm for filtering
rpkms_for_filter <- rpkms_mouse %>%
  as.data.frame() %>%
  rownames_to_column("Gene") %>%
  left_join(rpkms_human %>% as.data.frame() %>% rownames_to_column("Gene"),
            by = "Gene", suffix = c("_mouse", "_human")) %>%
  mutate(Min = do.call(pmin, select_if(., is.numeric))) %>%
  select(Gene, rpkm_min = Min)
```


Find non-significant (p > 0.2) genes that are well expressed (baseMean > 400, log(rpkm) > 4).

```{r message = FALSE}
# define cutoffs
p_cutoff <- 0.2
bm_cutoff <- 400
rpkm_cutoff <- 4

# join DE table with minimum rpkm table for each gene
# filter based on cutoffs
# add column with human gene name
non_sig <- de %>%
  left_join(rpkms_for_filter, by = "Gene") %>%
  filter(padj > p_cutoff, 
         baseMean > bm_cutoff,
         rpkm_min > rpkm_cutoff) %>%
  mutate(Gene_human = convertIDs(Gene))

non_sig_genes <- non_sig %>%
  select(Mouse = Gene, Human = Gene_human)

# output
write_tsv(
  non_sig_genes, 
  "../output/figure_3_s3_species_de_mouse_human_unchanged.txt"
  )
```


### Merge gene lists

```{r message = FALSE}
non_sig <- read_tsv("../output/figure_3_s3_species_de_mouse_human_unchanged.txt", 
                    col_names = c("Gene_mouse", "Gene_human"), skip = 1)
genes <- non_sig %>%
  select(Gene = "Gene_human") %>%
  mutate(Type = "HK", Cell_type = "HK") %>%
  rbind(se_genes)
```

## Single nuclei human

### Prepare single nuclei data

#### Load counts and derive metadata

Import single nucle count matrix (download count matrix from GEO - GSE97930)

```{r message = FALSE}
cb_dge <- fread("../data/GSE97930_CerebellarHem_snDrop-seq_UMI_Count_Matrix_08-01-2017.txt") %>%
  as.data.frame() %>%
  column_to_rownames("Gene") %>%
  as.matrix()
```

Derive metadata from sample names. Remove underscores in cell type name for Seurat. Rename cell types to match convention in this paper. Also, change the incorrectly named populations Purk1 and Purk2 to Basket and Int2 (interneuron population 2)

```{r}
lookup <- c(
  "Gran" = "Granule",
  "Purk1" = "Basket",
  "Purk2" = "Int2",
  "End" = "End",
  "Per" = "Perc",
  "Ast" = "Astrocyte",
  "Ast.cer" = "Bergmann.glia",
  "Oli" = "Oligo",
  "OPC" = "OPC",
  "OPC.cer" = "OPC.cer",
  "Mic" = "Microglia"
)

meta <- tibble(
  Orig_name = colnames(cb_dge),
  Sample = gsub("_Cer", ".cer", Orig_name)
) %>%
  separate(Sample, into = c("Cell_type", "Batch", "Barcode"), sep = "_") %>%
  mutate(Cell_type = lookup[Cell_type]) %>%
  mutate(Sample = paste(Cell_type, Batch, Barcode, sep = "_"))
```

Counts for each cell type

```{r}
meta %>%
  group_by(Cell_type) %>%
  count()
```



Initliaze Seurat object. Keep all genes expressed in at least 3 cells (~0.1% of data). Keep all cells with at least 200 detected genes. Add batch to Seurat metadata.

```{r warning = FALSE}
colnames(cb_dge) <- meta$Sample
cb_sn <- CreateSeuratObject(
  raw.data = cb_dge, 
  min.cells = 3, 
  min.genes = 200, 
  project = "cb_sn"
  )
cb_sn <- AddMetaData(
  object = cb_sn, 
  metadata = meta %>% column_to_rownames("Sample"), 
  col.name = "Batch"
  )
```

Remove cells that have fewer than 300 genes or more than 5,000 genes (outliers)

```{r}
cb_sn <- FilterCells(object = cb_sn, subset.names = "nGene", 
                     low.thresholds = 300, high.thresholds = 5000)
```


#### Normalizing the data

Use global scaling to normalize the gene expression measurements for each cell by the total expression, multiply this by a scale factor (10,000 by default), and log-transform the result.

```{r}
cb_sn <- NormalizeData(object = cb_sn, 
                       normalization.method = "LogNormalize", 
                       scale.factor = 10000)
```



### Plot gene expression from single nuclei data

#### Purk1 and Purk2 contain interneurons and not Purkinje neurons


Canonical Purkinje markers

```{r}
VlnPlot(object = cb_sn, x.lab.rot = TRUE,
        features.plot = c("CALB1", "CA8"))
ggsaveToPdf("../figures/figure_3_s3_human_markers_purkinje.pdf", h = 4, w = 8)
```



Cerebellar interneuron markers

```{r}
VlnPlot(object = cb_sn, x.lab.rot = TRUE,
        features.plot = c("SLC6A1", "TFAP2B"))
ggsaveToPdf("../figures/figure_3_s3_human_markers_interneuron.pdf", h = 4, w = 8)
```



Cerebellar molecular layer markers

```{r}
VlnPlot(object = cb_sn, x.lab.rot = TRUE,
        features.plot = c("SORCS3", "LYPD6"))
ggsaveToPdf("../figures/figure_3_s3_human_markers_basket.pdf", h = 4, w = 8)
```



#### Expresion of granule cell marker, human-enriched, and mouse-enriched genes

Granule cell markers

```{r}
VlnPlot(object = cb_sn, x.lab.rot = TRUE,
        features.plot = c("FAT2", "RBFOX3"))
ggsaveToPdf("../figures/figure_3_s3_markers_granule.pdf", h = 4, w = 8)
```


Granule cell human-enriched genes

```{r}
VlnPlot(object = cb_sn, x.lab.rot = TRUE,
        features.plot = c("VWC2", "CCDC175"))
ggsaveToPdf("../figures/figure_3_s3_markers_granule_human_enr.pdf", h = 4, w = 8)
```


Granule cell mouse-enriched genes

```{r}
VlnPlot(object = cb_sn, x.lab.rot = TRUE,
        features.plot = c("CNKSR3", "ECE1"))
ggsaveToPdf("../figures/figure_3_s3_markers_granule_mouse_enr.pdf", h = 4, w = 8)
```



#### Distribution of gene expression for each class of genes

For each cell type, plot distribution of gene expression for housekeeping, human-enriched, and mouse-enriched genes. From violin plots, looks like even marker genes have 0 expression in most samples. For summary stats, use either mean or % of cells that express gene over threshold of 1. Because our astrocyte population contains both Bergmann glia (cerebellar astrocytes) and conventional astrocytes, look at expression of these genes in both single nulcei populations. Use Kolmogorov-Smirnov test to determine whether differences in distribution are significant.


```{r warning = FALSE, message = FALSE}
comparisons <- tibble(
  Cell_type_SE = c("Granule", "Basket", "Astrocyte", "Astrocyte", "Oligo", "OPC", "OPC"),
  Cell_type_SN = c("Granule", "Basket", "Astrocyte", "Bergmann.glia", "Oligo", "OPC", "OPC.cer")
)

plotDist <- function(x, y) {
  # subset for housekeeping genes or species-enriched genes
  genes_sub <- genes %>%
    filter(Cell_type %in% c(y, "HK"))
  
  # get normalized counts for gene subset - reshape
  norm_counts <- cb_sn@data[rownames(cb_sn@data) %in% genes_sub$Gene, ] %>%
    as.matrix() %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column("Sample") %>% 
    left_join(meta, by = "Sample") %>%
    melt() %>%
    rename(Gene = variable) %>%
    filter(Cell_type == x) %>%
    left_join(genes_sub %>% select(-Cell_type), by = "Gene")
  
  # mean
  gene_sum <- norm_counts %>%
    group_by(Gene, Type) %>%
    summarise(Mean = mean(value))
  
  # plot cdf for mean
  plotCDF(gene_sum, "Mean", "Type", c("grey50", "red", "blue"), 
          paste(x, " - Mean", sep = ""),
          paste("../figures/figure_3_s3_human_cdf_mean_", tolower(x), ".pdf", sep = ""))
  
  
  # for each gene, % of nuclei with expression above 1
  gene_sum_exp <- norm_counts %>%
    group_by(Gene, Type, value > 1) %>%
    count() %>%
    spread("value > 1", n) %>%
    rename("Not_exp" = "FALSE", "Exp" = "TRUE") %>%
    replace_na(list(Not_exp = 0, Exp = 0)) %>%
    mutate(Total = Not_exp + Exp) %>%
    mutate(Prop_exp = round(Exp / Total, 2))
  
  # plot proportion expressed
  plotCDF(gene_sum_exp, "Prop_exp", "Type", c("grey50", "red", "blue"), 
          paste(x, " - Proportion Expressed", sep = ""),
          paste("../figures/figure_3_s3_human_cdf_prop_exp_", tolower(x), ".pdf", sep = ""))
  
  
  # Kolmogorov-Smirnov Tests
  ks_pvals <- c(
        ks.test(gene_sum %>% filter(Type == "Human") %>% pull(Mean),
                gene_sum %>% filter(Type == "Mouse") %>% pull(Mean))$p.value,
        ks.test(gene_sum_exp %>% filter(Type == "Human") %>% pull(Prop_exp),
                gene_sum_exp %>% filter(Type == "Mouse") %>% pull(Prop_exp))$p.value,
        ks.test(gene_sum %>% filter(Type == "Human") %>% pull(Mean),
                gene_sum %>% filter(Type == "HK") %>% pull(Mean))$p.value,
        ks.test(gene_sum_exp %>% filter(Type == "Human") %>% pull(Prop_exp),
                gene_sum_exp %>% filter(Type == "HK") %>% pull(Prop_exp))$p.value
        )
  
  return( formatC(ks_pvals, format = "e", digits = 2) )
}

ks_all_stats <- mapply(plotDist, comparisons$Cell_type_SN, comparisons$Cell_type_SE)
rownames(ks_all_stats) <- c("Human_Mouse_mean", 
                            "Human_Mouse_prop_exp", 
                            "Human_HK_mean", 
                            "Human_HK_prop_exp")
ks_all_stats %>% 
  as.data.frame() %>%
  rownames_to_column("Comparison")
```







## Single cell mouse

### Prepare single cell data

#### Load counts and format metadata

Import single nucle count matrix and metadata (download count matrix and metadata from http://dropviz.org/)
```{r message = FALSE}
dge <- loadSparseDge("../data/F_GRCm38.81.P60Cerebellum_ALT.raw.dge.txt.gz")

outcomes <- read_rds("../data/F_GRCm38.81.P60Cerebellum_ALT.cell_cluster_outcomes.RDS")
annot <- read_rds("../data/annotation.BrainCellAtlas_Saunders_version_2018.04.01.RDS")
```


Using metadata, exclude cells of class doublet, min_genes, and outlier. Match cluster / subcluster to cell type and rename using conventions in this paper.
```{r}
annot_cb <- annot %>%
  filter(tissue == "CB") %>%
  mutate(Cell_type = c(
    "Granule", "End.stalk.1", "End.stalk.2", "End.stalk.3", 
    "End.tip.1", "End.tip.2", "Mural.1", "Mural.2", "Mural.3",
    "Purkinje", "Basket.1", "Basket.2", "Basket.3", "Basket.4",
    "Uni.brush", "Golgi", "IC", "Microglia", "Oligo.1", "Oligo.2",
    "OPC", "Bergmann.glia", "Astrocyte.1", "Astrocyte.2", "CP"
  ))

meta_ms <- outcomes %>%
  rownames_to_column("Sample") %>%
  mutate_all(as.character) %>%
  mutate(sub = ifelse(subcluster == cluster,
                      paste(cluster, "1", sep = "-"),
                      subcluster)) %>%
  replace_na(list("reason" = "cluster")) %>%
  filter(!(reason %in% c("doublet", "min_genes", "outlier"))) %>%
  left_join(annot_cb, by = c("sub" = "subcluster")) %>%
  mutate(Name = paste(Cell_type, Sample, sep = "_")) %>%
  separate(Sample, into = c("Batch", "Barcode"), sep = "_", remove = FALSE)

# Filter DGE and rename samples with cell type name
dge_filt <- dge[, meta_ms$Sample]
colnames(dge_filt) <- meta_ms$Name
```


Counts for each cell type

```{r}
meta_ms %>%
  group_by(Cell_type) %>%
  count()
```


Initliaze Seurat object. Keep all genes expressed in at least 3 cells. Keep all cells with at least 200 detected genes. Add batch to Seurat metadata.

```{r warning = FALSE}
cb_sn_ms <- CreateSeuratObject(raw.data = dge_filt, min.cells = 3, 
                               min.genes = 200, project = "sn_cb_mouse")
cb_sn_ms <- AddMetaData(
  object = cb_sn_ms, 
  metadata = meta_ms %>% column_to_rownames("Name"), 
  col.name = "Batch"
  )
```

Remove cells that have fewer than 300 genes or more than 5,000 genes (outliers)

```{r}
cb_sn_ms <- FilterCells(object = cb_sn_ms, subset.names = "nGene", 
                        low.thresholds = 300, high.thresholds = 5000)
```


#### Normalizing the data

Use global scaling to normalize the gene expression measurements for each cell by the total expression, multiply this by a scale factor (10,000 by default), and log-transform the result.

```{r}
cb_sn_ms <- NormalizeData(object = cb_sn_ms, 
                          normalization.method = "LogNormalize", 
                          scale.factor = 10000)
```



### Plot gene expression from single cell data

#### Check markers for Purkinje and interneuron clusters.


Canonical Purkinje markers

```{r}
VlnPlot(object = cb_sn_ms, x.lab.rot = TRUE, size.x.use = 8,
        features.plot = c("Calb1", "Car8"))
ggsaveToPdf("../figures/figure_3_s3_mouse_markers_purkinje.pdf", h = 3, w = 8)
```



Cerebellar interneuron markers

```{r}
VlnPlot(object = cb_sn_ms, x.lab.rot = TRUE, size.x.use = 8,
        features.plot = c("Slc6a1", "Tfap2b"))
ggsaveToPdf("../figures/figure_3_s3_mouse_markers_interneuron.pdf", h = 3, w = 8)
```



Cerebellar molecular layer markers

```{r}
VlnPlot(object = cb_sn_ms, x.lab.rot = TRUE, size.x.use = 8,
        features.plot = c("Sorcs3", "Lypd6"))
ggsaveToPdf("../figures/figure_3_s3_mouse_markers_basket.pdf", h = 3, w = 8)
```



#### Expresion of granule cell marker, human-enriched, and mouse-enriched genes

Granule cell markers

```{r}
VlnPlot(object = cb_sn_ms, x.lab.rot = TRUE, size.x.use = 8,
        features.plot = c("Fat2", "Rbfox3"))
ggsaveToPdf("../figures/figure_3_s3_mouse_markers_granule.pdf", h = 3, w = 8)
```


Granule cell human-enriched genes

```{r}
VlnPlot(object = cb_sn_ms, x.lab.rot = TRUE, size.x.use = 8,
        features.plot = c("Vwc2", "Ccdc175"))
ggsaveToPdf("../figures/figure_3_s3_mouse_markers_granule_human_enr.pdf", h = 3, w = 8)
```


Granule cell mouse-enriched genes

```{r}
VlnPlot(object = cb_sn_ms, x.lab.rot = TRUE, size.x.use = 8,
        features.plot = c("Cnksr3", "Ece1"))
ggsaveToPdf("../figures/figure_3_s3_mouse_markers_granule_mouse_enr.pdf", h = 3, w = 8)
```



#### Distribution of gene expression for each class of genes

For each cell type, plot distribution of gene expression for housekeeping, human-enriched, and mouse-enriched genes. From violin plots, looks like even marker genes have 0 expression in most samples. For summary stats, use either mean or % of cells that express gene over threshold of 1. Because our astrocyte population contains both Bergmann glia (cerebellar astrocytes) and conventional astrocytes, look at expression of these genes in both single nulcei populations. Use Kolmogorov-Smirnov test to determine whether differences in distribution are significant.


```{r warning = FALSE, message = FALSE}
comparisons <- tibble(
  Cell_type_SE = c("Granule", rep("Basket", 4), rep("Astrocyte", 3), 
                   rep("Oligo", 2), "OPC"),
  Cell_type_SN = c("Granule", "Basket.1", "Basket.2", "Basket.3", "Basket.4", 
                   "Astrocyte.1", "Astrocyte.2", "Bergmann.glia", "Oligo.1", 
                   "Oligo.2", "OPC")
)

plotDist <- function(x, y) {
  # subset for housekeeping genes or species-enriched genes
  genes_sub <- genes %>%
    filter(Cell_type %in% c(y, "HK")) %>%
    mutate(Gene = convertIDs(Gene, from = "human", to = "mouse"))
  
  # get normalized counts for gene subset - reshape
  norm_counts <- cb_sn_ms@data[rownames(cb_sn_ms@data) %in% genes_sub$Gene, ] %>%
    as.matrix() %>%
    t() %>%
    as.data.frame() %>%
    rownames_to_column("Name") %>% 
    left_join(meta_ms, by = "Name") %>%
    melt() %>%
    rename(Gene = variable) %>%
    filter(Cell_type == x) %>%
    left_join(genes_sub %>% select(-Cell_type), by = "Gene")
  
  # mean
  gene_sum <- norm_counts %>%
    group_by(Gene, Type) %>%
    summarise(Mean = mean(value))
  
  # plot cdf for mean
  plotCDF(gene_sum, "Mean", "Type", c("grey50", "red", "blue"), 
          paste(x, " - Mean", sep = ""),
          paste("../figures/figure_3_s3_mouse_cdf_mean_", tolower(x), ".pdf", sep = ""))
  
  
  # for each gene, % of nuclei with expression above 1
  gene_sum_exp <- norm_counts %>%
    group_by(Gene, Type, value > 1) %>%
    count() %>%
    spread("value > 1", n) %>%
    rename("Not_exp" = "FALSE", "Exp" = "TRUE") %>%
    replace_na(list(Not_exp = 0, Exp = 0)) %>%
    mutate(Total = Not_exp + Exp) %>%
    mutate(Prop_exp = round(Exp / Total, 2))
  
  # plot proportion expressed
  plotCDF(gene_sum_exp, "Prop_exp", "Type", c("grey50", "red", "blue"), 
          paste(x, " - Proportion Expressed", sep = ""),
          paste("../figures/figure_3_s3_mouse_cdf_prop_exp_", tolower(x), ".pdf", sep = ""))
  
  
  # Kolmogorov-Smirnov Tests
  ks_pvals <- c(
        ks.test(gene_sum %>% filter(Type == "Human") %>% pull(Mean),
                gene_sum %>% filter(Type == "Mouse") %>% pull(Mean))$p.value,
        ks.test(gene_sum_exp %>% filter(Type == "Human") %>% pull(Prop_exp),
                gene_sum_exp %>% filter(Type == "Mouse") %>% pull(Prop_exp))$p.value,
        ks.test(gene_sum %>% filter(Type == "Human") %>% pull(Mean),
                gene_sum %>% filter(Type == "HK") %>% pull(Mean))$p.value,
        ks.test(gene_sum_exp %>% filter(Type == "Human") %>% pull(Prop_exp),
                gene_sum_exp %>% filter(Type == "HK") %>% pull(Prop_exp))$p.value
        )
  
  return( formatC(ks_pvals, format = "e", digits = 2) )
}

ks_all_stats <- mapply(plotDist, comparisons$Cell_type_SN, comparisons$Cell_type_SE)
rownames(ks_all_stats) <- c("Human_Mouse_mean", "Human_Mouse_prop_exp", "Human_HK_mean", "Human_HK_prop_exp")
ks_all_stats %>% 
  as.data.frame() %>%
  rownames_to_column("Comparison")
```
