---
title: "R code for Figure 2 - figure supplement 1 - Heatmap of rat top SI genes and RNA-seq metrics and qualtiy control"
author: "Xiao Xu (xxu@rockefeller.edu)"
date: "June 2018"
output: 
  html_document:
    toc: yes
  html_notebook:
    toc: yes
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. This notebook contains R code for figures and analyses related to Figure 2 - figure supplement 1 in the manuscript:

Xu X, Stoyanova EI, Lemiesz A, Xing J, Mash DC, Heintz N (2018). Species and Cell-Type Properties of Classically Defined Human and Rodent Neurons and Glia.

Specifically, this workflow will generate the heatmap of rat top SI gene expression, as well as the RNA-seq metrics and quality control figures in Figure 2 - figure supplement 1.

Load libraries and functions:

```{r warning = FALSE, message = FALSE}
# source custom functions
source("functions.R")

# libraries for gene expression analysis
library(DESeq2)

# libraries for heatmaps and colors
library(RColorBrewer)
library(gplots)

# data manipulation
library(plyr)
library(reshape2)

# work in the tidyverse
library(tidyverse)
library(ggbeeswarm)
```



  
# Plot SI heatmap for rat

Read in counts and length data and calculate rpkms:

```{r message = FALSE, warning = FALSE}
# read in table of counts
# exclude unsorted sample for SI calculations
counts <- read_tsv("../data/rat_all_wg_counts.txt") %>%
  select( -contains("_uns") )

# create metadata for samples
meta <- tibble(
  Sample = colnames(counts)[-1],
  Cell_type = rep(
    c("Granule", "Purkinje", "Basket", "Astrocyte", "Oligo", "OPC"),
    each = 2
    )
  )

# read in gene lengths
lengths <- read_tsv("../data/ref/rat_ensembl_wg_lengths.txt", 
                    col_names = c("Gene", "Length"))

# calculate rpkms
rpkms <- getNormCounts(counts, lengths, type = "RPKM")
```


Calculate SI:
```{r eval = FALSE}
set.seed(1)
si_table <- siWithReps(rpkms %>% as.data.frame(), 
                       meta$Cell_type, 
                       bottom = 0, reps = TRUE, iterations = 1000,
                       parallel = TRUE, cores = 32)
write_tsv(si_table %>% 
            as.data.frame() %>%
            rownames_to_column("Gene"), 
          "../output/si_rat.txt")
```


Read in specificity index data, identify the top 20 genes for each cell type, and get rpkms for each gene across all conditions:

```{r message = FALSE}
# read in specificity index table
si_table <- read_tsv("../output/si_rat.txt") %>% 
  column_to_rownames("Gene")

# get top 20 si genes for each cell type
si_top20 <- as.vector( apply(si_table, 2, sortedIndex, table = si_table, top = 20) )
# get rpkms for top si genes
rpkms_si <- expFromGeneList(rpkms, si_top20)
# convert rpkms to z scores for heatmap plotting
mat <- as.matrix( apply(rpkms_si, 1, getZScores) )
```


Specify heatmap palette and sidebar colors

```{r}
# palette for z-scores
cor_col <- rev( colorRampPalette( brewer.pal(9, 'RdBu') )(100) )
cor_breaks <- seq(-2.5, 2.5, length.out = 101)
# sidebar for heatmap.2
cell_type_cols <- c(
  Granule = "#33A02C",
  Purkinje = "#E31A1C",
  Basket = "#FF6600",
  Astrocyte = "#0033FF",
  Oligo = "#00CCCC",
  OPC = "#00FFFF"
)
sidebar_row <- rep(cell_type_cols, each = 2)
sidebar_col <- rep(cell_type_cols, each = 20)
```


Final heatmap:

```{r warning=FALSE}
saveToPdf("../figures/figure_2_s1_heatmap_rat_si.pdf")
hm <- heatmap.2(
  mat,
  dendrogram = "none", Rowv = FALSE, Colv = FALSE, # no dendrogram, no reordering
  col = cor_col, breaks = cor_breaks,  # colors for values
  RowSideColors = sidebar_row, ColSideColors = sidebar_col,  # sidebar
  scale = "none", trace = "none",
  key.title = NA, key.ylab = NA,
  cexRow = 1, cexCol = 1, margins = c(2, 2),
  labCol = FALSE, labRow = FALSE,
  keysize = 1, key.par = list(mar = c(4, 2, 1, 1))
)
dev.off()

# and display in this file
eval(hm$call)
```




## Heatmap to accompany IGV tracks for human


```{r message = FALSE}
# read in table of counts
counts <- read_tsv("../data/human_xk_pk_wg_counts.txt")
```

Create metadata for samples:

```{r}
meta <- tibble(
  Sample = colnames(counts)[-1],
  Cell_type = rep(
    c("Granule", "Basket", "Astrocyte", "Oligo", "OPC"),
    each = 2
    ),
  Individual = rep(c("XK", "PK"), 5)
  )
```



Normalize counts by length using FPKM (RPKM) calculation:

```{r message = FALSE}
# read in gene lengths
lengths <- read_tsv("../data/ref/hg38_refseq_wg_longest.txt", 
                    col_names = c("Gene", "Length"))

# calculate rpkms
rpkms <- getNormCounts(counts, lengths, type = "RPKM")
```

Genes to visualize

```{r}
igv_genes <- c("RBFOX3", "FAT2", "SORCS3", "MARCH11", "ALDH1L1", "SLC1A3",
               "OLIG2", "MOG", "PDGFRA")
rpkms_for_igv <- rpkms[igv_genes, ] %>% t()
```

Specify colors for heatmap:

```{r}
# palette for rpkms
cor_col <- colorRampPalette( brewer.pal(9, 'YlOrRd') )(100)
cor_breaks <- seq(-2, 6, length.out = 101)

# sidebar for heatmap.2
cell_type_cols <- c(
  Granule = "#33A02C",
  Basket = "#FF6600",
  Astrocyte = "#0033FF",
  Oligo = "#00CCCC",
  OPC = "#00FFFF"
)
sidebar_row <- rep(cell_type_cols, each = 2)
sidebar_col <- rep(cell_type_cols, c(2, 2, 2, 2, 1))
```

Plot heatmap:

```{r warning=FALSE}
pdf("../figures/figure_2_s1_heatmap_human_for_igv.pdf", width = 11, height = 8.5)
hm <- heatmap.2(
  rpkms_for_igv,
  dendrogram = "none", Rowv = FALSE, Colv = FALSE, # no dendrogram, no reordering
  col = cor_col, breaks = cor_breaks,  # colors for values
  RowSideColors = sidebar_row, ColSideColors = sidebar_col,  # sidebar
  scale = "none", trace = "none",
  key.title = NA, key.ylab = NA,
  cexRow = 1, cexCol = 1, margins = c(2, 2),
  labCol = FALSE, labRow = FALSE, symkey = F,
  keysize = 1, key.par = list(mar = c(4, 2, 1, 1))
)
dev.off()

# and display in this file
eval(hm$call)
```


## Quality control

Read in table of picard metrics
```{r message = FALSE}
# specify order of samples for plotting
sample_order <- c("mouse.gfp", "mouse.a", "rat.a", "human.a", "cytoplasmic")
metrics <- read_tsv("../data/rnaseq_metrics.txt") %>%
  mutate(Sample = factor(Sample, levels = sample_order))
```



Violin plots of metrics
```{r}
lapply(
  c("Ribosomal", "Coding", "UTR", "Intronic", "Intergenic", "GC"),
  function(x) {
    p <- plotViolinMetrics(metrics, type = x)
    ggsaveToPdf(paste("../figures/figure_2_s1_metrics_", tolower(x), ".pdf"), 
                plot = p)
    p
  }
)
```


Summarized stacked bar chart

```{r}
metrics_mean <- metrics %>%
  group_by(Sample) %>%
  summarise(Ribosomal = mean(Ribosomal), 
            Coding = mean(Coding),
            UTR = mean(UTR),
            Intronic = mean(Intronic),
            Intergenic = mean(Intergenic)) %>%
  gather(-Sample, key = Metric, value = Value) %>%
  mutate(Sample = factor(Sample, levels = sample_order)) %>%
  mutate(Metric = factor(Metric, levels = c("Ribosomal", "Coding", "UTR", "Intronic", "Intergenic")))


ggplot(data = metrics_mean, aes(x = Sample, y = Value)) + 
  geom_bar(aes(fill = Metric), 
           stat = "identity",
           position = position_stack(reverse = TRUE)) +
  scale_fill_manual(values = brewer.pal(5, "Set2")) +
  theme_bw()
ggsaveToPdf("../figures/figure_2_s1_metrics_mean.pdf")
```



## PCA analysis of batch effects

### Human

Read counts and create metadata for samples

```{r message = FALSE}
counts <- read_tsv("../data/counts_for_batch_human.txt")

meta <- tibble(
  Sample = colnames(counts)[-1],
  Cell_type = rep( c("granule", "basket", "glia"), 4 ),
  Donor = rep( c("LT", "ZH", "SG", "OR"), each = 3),
  Batch = rep( c("2", "1"), each = 6),
  Gender = rep( c(rep("male", 3), rep("female", 3)), 2 )
)
```


Use deseq2 to generate normalized counts

```{r warning = FALSE, messsage = FALSE}
dds <- DESeqDataSetFromMatrix(counts %>% column_to_rownames("Gene"), 
                              meta %>% column_to_rownames("Sample"), 
                              design =~ Cell_type)
dds <- DESeq(dds)

# get normalized counts
rld <- rlogTransformation(dds, blind = TRUE)
```


Perform PCA

```{r}
pca <- calcPCA(rld, 
               intgroup = c("Donor", "Batch", "Gender", "Cell_type"),
               ntop = 250)
```


Plot importance of each principle component
```{r}
plotPCImportance(pca)
ggsaveToPdf("../figures/figure_2_s1_pca_human_importance.pdf")
```

Define function for custom plotting PCA

```{r}
plotPCACustom <- function(pca, x, y) {
  ggplot(pca, aes_string(x = x, y = y)) + 
    geom_point(aes(fill = Cell_type, shape = Batch, color = Gender),
               size = 6, stroke = 2, alpha = 0.8) +
    scale_color_manual(values = c("black", "red")) +
    scale_fill_manual(values = c("orange", "blue", "green", "red")) +
    scale_shape_manual(values = c(21, 22, 23, 24, 25)) +
    theme_bw()
}
```

Plot PCs 1 - 3 to check for batch effects

```{r}
plotPCACustom(pca, "PC1", "PC2")
ggsaveToPdf("../figures/figure_2_s1_pca_human_pc1_pc2.pdf", w = 6, h = 4)
plotPCACustom(pca, "PC1", "PC3")
ggsaveToPdf("../figures/figure_2_s1_pca_human_pc1_pc3.pdf", w = 6, h = 4)
```




### Mouse

```{r message = FALSE}
counts <- read_tsv("../data/counts_for_batch_mouse.txt")

meta <- tibble(
  Sample = colnames(counts)[-1],
  Cell_type = c( rep("granule", 3), rep("basket", 3), 
                 rep("astrocyte", 2), rep("purkinje", 5) ),
  Batch = as.factor( c(2, 2, 3, 2, 2, 1, 3, 3, 4, 4, 1, 5, 5) ),
  Gender = c("male", "male", "female", "male", "male", 
             "female", "female", "female", "female", 
             "female", "male", "female", "female")
)
```

Use deseq2 to generate normalized counts

```{r warning = FALSE, message = FALSE}
dds <- DESeqDataSetFromMatrix(counts %>% column_to_rownames("Gene"), 
                              meta %>% column_to_rownames("Sample"), 
                              design =~ Cell_type)
dds <- DESeq(dds)

# get normalized counts
rld <- rlogTransformation(dds, blind = TRUE)
```

Perform PCA

```{r}
pca <- calcPCA(rld, 
               intgroup = c("Batch", "Gender", "Cell_type"),
               ntop = 250)
```


Plot importance of each principle component

```{r}
plotPCImportance(pca)
ggsaveToPdf("../figures/figure_2_s1_pca_mouse_importance.pdf")
```


Plot PCs 1 - 3 to check for batch effects

```{r}
plotPCACustom(pca, "PC1", "PC2")
ggsaveToPdf("../figures/figure_2_s1_pca_mouse_pc1_pc2.pdf", w = 6, h = 4)
plotPCACustom(pca, "PC1", "PC3")
ggsaveToPdf("../figures/figure_2_s1_pca_mouse_pc1_pc3.pdf", w = 6, h = 4)
```



